src/test/ui/better_divergence_checking.rs

```diff
@@ -0,0 +1,35 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(exhaustive_patterns)]
+#![feature(never_type)]
+#![deny(unreachable_code)]
+
+pub enum Void {}
+
+pub fn uninhabited_parameter_i(_v: Void) {
+    // A function with an uninhabited parameter
+    // is permitted if its body is empty.
+}
+
+pub fn uninhabited_parameter_ii(v: !) -> i32 {
+    // A function with an uninhabited parameter
+    // is permitted if it simply returns a value
+    // as a trailing expression to satisfy the
+    // return type.
+    v
```

Comment:
I guess this answers my question to "how should the user write this code" -- instead of `match void { }` they would simply write `void`?

But [that will only work if the uninhabited parameter has type `!`...?](https://play.rust-lang.org/?gist=154ed08d778627693a4cef2d889860ec&version=nightly&mode=debug&edition=2015) Or, maybe we changed that in this PR somewhere?